flare/gp_from_aimd.py

```diff
@@ -0,0 +1,232 @@
+                 skip: int = 0,
+                 calculate_energy: bool = False,
+                 output_name: str = 'gp_from_aimd.out',
+                 max_atoms_from_frame: int = 0, max_trains: int = 10,
+                 min_atoms_added: int = 1,
+                 n_cpus: int = 1, shuffle_frames: bool = False,
+                 verbose: int = 0, model_write: str = ''):
+        """
+        Class which trains a GP off of an AIMD trajectory, and generates
+        error statistics between the DFT and GP calls.
+
+        :param gp:
+        :param rel_std_tolerance:
+        :param abs_std_tolerance:
+        :param parallel:
+        :param skip:
+        :param calculate_energy:
+        :param output_name:
+        :param max_atoms_from_frame: Largest # of atoms added from one frame
+        :param min_atoms_added: Only train when this many atoms have been added
+        :param freeze_hyps:
+        :param rescale_steps:
+        :param rescale_temps:
+        :param n_cpus:
+        """
+
+        self.frames = frames
+        if shuffle_frames:
+            np.random.shuffle(frames)
+        self.gp = gp
+        self.rel_std_tolerance = rel_std_tolerance
+        self.abs_std_tolerance = abs_std_tolerance
+        self.skip = skip
+        self.max_trains = max_trains
+        self.curr_step = 0
+        self.max_atoms_from_frame = max_atoms_from_frame
+        self.min_atoms_added = min_atoms_added
+        self.verbose = verbose
+        self.train_count = 0
+
+        self.parallel = parallel
+        # set pred function
+        if parallel:
+            if calculate_energy:
+                self.pred_func = predict_on_structure_par_en
+            else:
+                self.pred_func = predict_on_structure_par
+        else:
+            if calculate_energy:
+                self.pred_func = predict_on_structure_en
+            else:
+                self.pred_func = predict_on_structure
+
+        self.output_name = output_name
+
+        # set number of cpus for parallelization
+        self.n_cpus = n_cpus
+
+        # To later be filled in using the time library
+        self.start_time = None
+        self.pickle_name = model_write
+
+    def pre_run(self):
+        """
+        Various tasks to set up the AIMD training before commencing
+        the run through the AIMD trajectory
+        :return:
+        """
+
+        output.write_header(self.gp.cutoffs,
+                            self.gp.kernel_name,
+                            self.gp.hyps,
+                            self.gp.algo,
+                            dt=0,
+                            Nsteps=len(self.frames),
+                            structure=self.frames[0],
+                            std_tolerance=(self.rel_std_tolerance,
+                                           self.abs_std_tolerance),
+                            output_name=self.output_name)
+
+        self.start_time = time.time()
+
+    def run(self):
```

Comment:
Down the line we'll want to add a number of options for training on a fixed AIMD set; might make sense to generalize the code a little now to make way for future implementations. Passing over each frame and adding environments with high uncertainty makes sense and is data efficient, but it's not the only way (or the easiest). We should also make it easy to train on equally spaced frames, user specified frames, etc. @nw13slx has been playing with different approaches --- it would be good to coordinate with her on this.

Reply:
These points are all totally valid. 

This comment touches on the tradeoff to be made here between how much power to cede to the user and how much to build in as functionality, which then results in a more complex training manager. I am pretty conflicted about this myself, and I'm not sure where the line should be. I would love to solicit input from you + @nw13slx + everyone else at a developer meeting about this.

It feels simultaneously like a huge cop out and not a cop out to say that the user can do some of this already. 

For example, if the typehint is followed, one can pass in a list of Structures sliced by e.g.
`input_list[::10]` (though that will not check the MAE against every frame in the trajectory, which might be desirable). An earlier version of the TrajectoryTrainer I had allowed 'seed structures' to be passed in, which would then be applied to the GP before the run through the entire trajectory began. I decided against leaving it in, hoping that the ability to add environments one-by-one into the GP added via the add_one_env method would enable the user to do this before putting their GP in.

On the other perspective, in which the above argument is a cop out, the whole point of TrajectoryTrainer is to automate labor and pre-processing. Musing about where 'the line' is, I think the features you request are probably on the side of inclusion. We can add a 'skip through' option and some version of 'seed frames' which can be passed to the TrajectoryTrainer. One can also do the frames pre-processing or GP pre-processing if they like, and skip the above two features anyway.